src/js/msp/MSPConnector.js

```diff
@@ -6,85 +6,105 @@ import MSP from "../msp";
+        this.onConnectCallback = undefined;
+        this.onTimeoutCallback = undefined;
+        this.onDisconnectCallback = undefined;
+    }
+
+    connect(port, baud, onConnectCallback, onTimeoutCallback, onFailureCallback) {
+
+        this.port = port;
+        this.baud = baud;
+        this.onConnectCallback = onConnectCallback;
+        this.onTimeoutCallback = onTimeoutCallback;
+        this.onFailureCallback = onFailureCallback;
+
+        this.connectHandler = (openInfo) => {
+            if (openInfo) {
+                const disconnectAndCleanup = () => {
+                    serial.disconnect((result) => {
+                        console.log('Disconnected', result);
+
+                        MSP.clearListeners();
+
+                        this.onTimeoutCallback();
+                    });
+
+                    MSP.disconnect_cleanup();
+                };
+
+                FC.resetState();
+
+                // disconnect after 10 seconds with error if we don't get IDENT data
+                GUI.timeout_add('msp_connector', function () {
+                    if (!CONFIGURATOR.connectionValid) {
+                        gui_log(i18n.getMessage('noConfigurationReceived'));
+
+                        disconnectAndCleanup();
+                    }
+                }, 10000);
+
+                function read_serial_adapter(e) {
+                    read_serial(e.detail.buffer);
+                }
 
-                MSP.disconnect_cleanup();
-            };
+                serial.removeEventListener('receive', read_serial_adapter);
+                serial.addEventListener('receive', read_serial_adapter);
 
-            FC.resetState();
+                const mspHelper = new MspHelper();
+                MSP.listen(mspHelper.process_data.bind(mspHelper));
 
-            // disconnect after 10 seconds with error if we don't get IDENT data
-            GUI.timeout_add('msp_connector', function () {
-                if (!CONFIGURATOR.connectionValid) {
-                    gui_log(i18n.getMessage('noConfigurationReceived'));
+                MSP.send_message(MSPCodes.MSP_API_VERSION, false, false, () => {
+                    CONFIGURATOR.connectionValid = true;
 
-                    disconnectAndCleanup();
-                }
-            }, 10000);
+                    GUI.timeout_remove('msp_connector');
+                    console.log('Connected');
 
-            serial.onReceive.addListener(read_serial);
+                    this.onConnectCallback();
+                });
+            } else {
+                gui_log(i18n.getMessage('serialPortOpenFail'));
+                this.onFailureCallback();
+            }
+        };
 
-            const mspHelper = new MspHelper();
-            MSP.listen(mspHelper.process_data.bind(mspHelper));
+        this.disconnectHandler = (detail) => {
+            console.log('Disconnected', detail);
+            MSP.clearListeners();
+            MSP.disconnect_cleanup();
+        };
```

Comment:
Similar, this should be another method in the class instead of `this.bind`

Reply:
Done.